sunpy/map/mapbase.py

```diff
@@ -929,6 +925,19 @@ def reference_date(self):
             self.date
         )
 
+    def _set_reference_date(self, date):
+        """
+        Set the reference date using the same priority as `.GenericMap.reference_date`.
+
+        If a source subclass overrides `.GenericMap.reference_date`, it should override
+        this private method as well.
+        """
+        for keyword in ['date-avg', 'date-obs', 'date-beg', 'date-end']:
```

Comment:
Rather than reimplementing all this boiler plate code for the getter and setter why don't we take a leaf out of the coordinate frame book and define a list of keys for the various properties on each source?

(I would be ok with that being a followup PR if we want to get this one out)

Reply:
I leave this up to @ayshih to decide.

Reply:
It gets trickier if the reference date is split across multiple keys.  Future PR!